mino/minows/rpc.go

```diff
@@ -0,0 +1,422 @@
+const pathStream = "/stream"
+
+// Packet encapsulates a message sent over the network streams.
+type Packet struct {
+	Source  []byte
+	Payload []byte
+}
+
+type envelope struct {
+	author mino.Address
+	msg    serde.Message
+	err    error
+}
+
+// RPC
+// - implements mino.RPC
+type rpc struct {
+	logger zerolog.Logger
+
+	uri     string
+	handler mino.Handler
+	mino    *minows
+	factory serde.Factory
+	context serde.Context
+}
+
+func (r rpc) Call(ctx context.Context, req serde.Message,
+	players mino.Players) (<-chan mino.Response, error) {
+	if players == nil || players.Len() == 0 {
+		resp := make(chan mino.Response)
+		close(resp)
+		return resp, nil
+	}
+
+	addrs, err := toAddresses(players)
+	if err != nil {
+		return nil, err
+	}
+
+	r.addPeers(addrs)
+
+	result := make(chan envelope, len(addrs))
+	for _, addr := range addrs {
+		if r.mino.myAddr.Equal(addr) {
+			request := mino.Request{Address: r.mino.myAddr, Message: req}
+			reply, err := r.handler.Process(request)
+			result <- envelope{r.mino.myAddr, reply, err}
+		} else {
+			go func(addr address) {
+				reply, err := r.unicast(ctx, addr, req)
+				result <- envelope{addr, reply, err}
+			}(addr)
+		}
+	}
+
+	responses := make(chan mino.Response, len(addrs))
+	go func() {
+		defer close(responses)
+		for i := 0; i < len(addrs); i++ {
+			select {
+			case <-ctx.Done():
+				return
+			case env := <-result:
+				if env.err != nil {
+					responses <- mino.NewResponseWithError(env.author, env.err)
+				} else {
+					responses <- mino.NewResponse(env.author, env.msg)
+				}
+			}
+		}
+	}()
+	return responses, nil
+}
+
+func (r rpc) Stream(ctx context.Context, players mino.Players) (mino.Sender, mino.Receiver, error) {
+	if players == nil || players.Len() == 0 {
+		return nil, nil, xerrors.New("no players")
+	}
+
+	initiator, err := libp2p.New(libp2p.NoListenAddrs)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("could not start host: %v", err)
+	}
+
+	go func() {
+		<-ctx.Done()
+		err := initiator.Close()
+		if err != nil {
+			r.logger.Error().Err(err).Msg("could not close host")
+		}
+	}()
+
+	errs := make(chan error, players.Len())
+	streams := make(chan network.Stream, players.Len())
+	var wg sync.WaitGroup
+	wg.Add(players.Len())
+	for iter := players.AddressIterator(); iter.HasNext(); {
+		player := iter.GetNext()
+		addr, ok := player.(address)
+		if !ok {
+			return nil, nil, xerrors.Errorf("%v: %T",
+				ErrWrongAddressType, player)
+		}
+		go func(addr address) {
+			defer wg.Done()
+			initiator.Peerstore().AddAddr(addr.identity, addr.location,
+				peerstore.PermanentAddrTTL)
+
+			stream, err := initiator.NewStream(ctx, addr.identity,
+				protocol.ID(r.uri+pathStream))
+			if err != nil {
+				errs <- xerrors.Errorf("could not open stream: %v", err)
+				return
+			}
+			streams <- stream
+			go func() {
+				<-ctx.Done()
+				err = stream.Reset()
+				if err != nil {
+					r.logger.Error().Err(err).Msg("could not reset stream")
+				}
+			}()
+		}(addr)
+	}
```

Comment:
a bit of vertical spacing owuldn't hurt readablity here :-) 

Also, would it make sense to have a top-level function to create a stream, instead of having five levels of indentation and two nested unnamed functions ?

Reply:
Yes, that part is awful. I looked at the different anonymous functions but in the end decided against putting them as private methods to `rpc`. Well, some of them were added, it was worse before :)

But as you don't like it either, I'll have another look to see if there are some easy refactorings there.